### ISSUE TITLE
[Clang++] Reference to inactive union member is not produce error in the constexpr context

### ISSUE THREAD

**Issue by safocl** [llvm/llvm-project#126792]
I believe the code here produces undefined behavior, but in the context of a contextual expression the compiler does not issue a compilation error.
https://godbolt.org/z/758Knbs8a (GCC and MSVC do the same)
```cpp
consteval int foo2(){
    union uu{int i; short ui;} u;
    // u.i = 11;         // with this behavior is defined
    int& i_ref = u.i; // u.i is not active union member (lifetime is not started)

    u.i = 11;

    return i_ref;
}

int main(){
  return foo2();
}
```
Related: https://invent.kde.org/plasma/kwin/-/merge_requests/7120

[\[class.union.general\]](https://eel.is/c++draft/class.union.general)

> A union is a class defined with the class-key union.
> 
> In a union, a non-static data member is active if its name refers to an object whose lifetime has begun and has not ended ([basic.life]). At most one of the non-static data members of an object of union type can be active at any time, that is, the value of at most one of the non-static data members can be stored in a union at any time.

[\[basic.memobj#intro.object-1.sentence-2\]](https://eel.is/c++draft/basic.memobj#intro.object-1.sentence-2)
> The constructs in a C++ program create, destroy, refer to, access, and manipulate objects. An object is created [...] when implicitly changing the active member of a union [...]

[\[class.union#general-3.sentence-2\]](https://eel.is/c++draft/class.union#general-3.sentence-2)
> Each non-static data member is allocated as if it were the sole member of a non-union class.

[\[basic.life#2.sentence-3\]](https://eel.is/c++draft/basic.life#2.sentence-3)
> except that if the object is a union member or subobject thereof, its lifetime only begins if that union member is the initialized member in the union ([dcl.init.aggr], [class.base.init]), or as described in [class.union], [class.copy.ctor], and [class.copy.assign], and except as described in [allocator.members].

[\[class.union#general-5.3\]](https://eel.is/c++draft/class.union#general-5.3)
> 5 When the left operand of an assignment operator involves a member
> access expression ([expr.ref]) that nominates a union member, it may
> begin the lifetime of that union member, as described below. For an
> expression E, define the set S(E) of subexpressions of E as follows:
> [...]
> (5.3) -- Otherwise, S(E) is empty.

and
https://eel.is/c++draft/class.union#general-example-2
_(that is, the members of the union do not become active (lifetime does not start for any of them) if the union is not declared with initialization, or when explicit or implicit initialization of one of its members has not occurred.)_

[\[dcl.ref#6.sentence-6\]](https://eel.is/c++draft/dcl.ref#6.sentence-6)
> The behavior of an evaluation of a reference ([expr.prim.id], [expr.ref]) that does not happen after ([intro.races]) the initialization of the reference is undefined.

And only follow behavior is defined:

[\[basic.life#10.sentence-1\]](https://eel.is/c++draft/basic.life#10.sentence-1)
> After the lifetime of an object has ended and before the storage which
> the object occupied is reused or released, if a new object is created
> at the storage location which the original object occupied and the original
> object was transparently replaceable by the new object, a pointer that
> pointed to the original object, a reference that referred to the original
> object, or the name of the original object will automatically refer to
> the new object and, once the lifetime of the new object has started,
> can be used to manipulate the new object.

[\[basic.life#8\]](https://eel.is/c++draft/basic.life#8)
> 8 Similarly, before the lifetime of an object has started but after the storage which the object will occupy has been allocated or, after the lifetime of an object has ended and before the storage which the object occupied is reused or released, any glvalue that refers to the original object may be used but only in limited ways. For an object under construction or destruction, see [class.cdtor]. Otherwise, such a glvalue refers to allocated storage ([basic.stc.dynamic.allocation]), and using the properties of the glvalue that do not depend on its value is well-defined. The program has undefined behavior if
> (8.1) -- the glvalue is used to access the object, or

[\[expr.post#expr.ref-2.sentence-1\]](https://eel.is/c++draft/expr.post#expr.ref-2.sentence-1)
> For the first option (dot), [...] if the id-expression names a non-static data member, the first expression shall be a glvalue.

[\[https://eel.is/c++draft/dcl.ref#example-3\]](https://eel.is/c++draft/dcl.ref#example-3)
> [Example 3: 
```cpp
int &f(int&);
int &g();
extern int &ir3;
int *ip = 0;
int &ir1 = *ip;                 // undefined behavior: null pointer
int &ir2 = f(ir3);              // undefined behavior: ir3 not yet initialized
int &ir3 = g();
int &ir4 = f(ir4);              // undefined behavior: ir4 used in its own initializer

char x alignas(int);
int &ir5 = *reinterpret_cast<int *>(&x);    // undefined behavior: initializer refers to char object
```
> — end example]

(_// undefined behavior: ir3 not yet initialized_)

_(that is, there are no exceptions to the fact that the reference must be initialized to a valid object. But it is initialized by a member of a union whose lifetime has not started.)_

**Comment by shafik** [comment]
I am sure attempting to change the active member via a reference is undefined but we seem to have implementation divergence there: https://godbolt.org/z/azxhzeojK

**Comment by shafik** [comment]
I had a quick chat w/ CWG folks, so [basic.life p8](https://eel.is/c++draft/basic.life#8) says that the reference binding is fine. We are not violating any of the bullets 8.1 to 8.4. So we have storage we can bind to it but we are limited in what we can do with that reference.

**Comment by frederick-vs-ja** [comment]
> I had a quick chat w/ CWG folks, so [basic.life p8](https://eel.is/c++draft/basic.life#8) says that the reference binding is fine. We are not violating any of the bullets 8.1 to 8.4. So we have storage we can bind to it but we are limited in what we can do with that reference.

Agreed. This is clarified by [CWG453](https://cplusplus.github.io/CWG/issues/453.html) recently.

Another example accepted by all implementations while its validity was unclear before resolving [CWG453](https://cplusplus.github.io/CWG/issues/453.html).

```C++
struct S {};
constexpr S s = s; // binding non-living `s` to the `const S&` parameter of the copy constructor of `S`
```

**Comment by safocl** [comment]
> but we seem to have implementation divergence there: https://godbolt.org/z/azxhzeojK

this turned out to be correct behavior for clang and incorrect for those compilers that passed it? (as I understand it)

**Comment by shafik** [comment]
> > but we seem to have implementation divergence there: https://godbolt.org/z/azxhzeojK
> 
> this turned out to be correct behavior for clang and incorrect for those compilers that passed it? (as I understand it)

Correct.

**Comment by safocl** [comment]
ok, I hope this discussion was useful and will be useful to others to clarify the meaning of the C++ standard
